File: services/url-shortener-backend-node-nestjs/src/link.utils.ts
const murmurhash = require("murmurhash")
const Hashids = require('hashids');
const hashids = new Hashids();
let SEQUENCE = 1;
export const generateId = (url: string) => {
    const hash: number = murmurhash.v3(url) & 0x7fffffff;
    const sequence = SEQUENCE++ % 1024
    const id = hash << 10 | sequence;
    return hashids.encode(id);
};
export const validUrl = (string: string) => {
    try {
        let url = new URL(string);
        return url.protocol.match("https?:")
    } catch (_) {
        return false;
    }
}

File: services/url-shortener-backend-java-spring/src/main/java/com/github/dungviettran89/urlshortener/utils/LinkUtils.java
package com.github.dungviettran89.urlshortener.utils;

import org.hashids.Hashids;

import java.net.URL;

@SuppressWarnings("UnstableApiUsage")
public class LinkUtils {
    private static Hashids hashids = new Hashids();
    private static long SEQUENCE = 0L;

    public static String generate(String url) {
        long hash = url.hashCode() & 0x7fffffff;
        long sequence = SEQUENCE++ % 1024;
        long id = hash << 10 | sequence;
        return hashids.encode(id);
    }

    public static boolean valid(String url) {
        try {
            new URL(url).toURI();
            return url.length() < 2048;
        } catch (Exception e) {
            return false;
        }
    }
}
